core/io/image.compat.inc

```diff
@@ -0,0 +1,41 @@
+/**************************************************************************/
+/*  image.compat.inc                                                      */
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef DISABLE_DEPRECATED
+
+int32_t Image::_get_mipmap_offset_bind_compat_94243(int p_mipmap) const {
+	return get_mipmap_offset(p_mipmap);
+}
+
+void Image::_bind_compatibility_methods() {
+	ClassDB::bind_compatibility_method(D_METHOD("get_mipmap_offset", "mipmap"), &Image::_get_mipmap_offset_bind_compat_94243);
+}
```

Comment:
@raulsntos @dsnopek Not sure how to deal with this compat method, it seems GDExtension and C# have conflicting needs.

- GDExtension (at least our CI checks) want a compat method as the return type changed from `int32_t` to `int64_t`.
- C# doesn't like it and complains:
```
/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/Image.cs(1207,17): error CS0111: Type 'Image' already defines a member called 'GetMipmapOffset' with the same parameter types [/home/runner/work/godot/godot/modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj]
```

Reply:
Hm. Having a compatibility method for the change in return type is the right thing to do for GDExtension. Perhaps the C# code generation hasn't yet had to take into account a compatibility method where the only change is the return type?

Reply:
The glue generation seems to have logic to skip methods with conflicting signature:

https://github.com/godotengine/godot/blob/293c0f76463a4e1c2c2e86d8fbfc3e04f491b347/modules/mono/editor/bindings_generator.cpp#L3928-L3935

But the implementation seems to ignore the return type, and I don't really understand the comment justifying it.

https://github.com/godotengine/godot/blob/293c0f76463a4e1c2c2e86d8fbfc3e04f491b347/modules/mono/editor/bindings_generator.cpp#L4762-L4765

Reply:
C# can't have overloads where the only thing that changes is the return type, so these kinds of changes break compat for C#.

The bindings generator avoids generating these compat methods, to avoid generating invalid C# that won't compile. That's what the comment is trying to say, we don't compare the return type because we don't care if they are different it should still return `true` in this case (the method is still considered conflicting even if the return type is different).

I'm pretty sure we already have cases of compat methods where the only thing that changes is the return type, so this should already be handled. Not sure why it would generate invalid C# in this case, it may have something to do with the type, `int32_t` and `int64_t` are really both `INT` with different metadata, so maybe ClassDB gets confused about this and it's not properly registering the compat method.

The bindings generator only checks conflicting methods of compat methods, so if it's not registered as a _compat method_ it won't check. We assume there are no conflicts in the non-compat methods because ClassDB doesn't support method overloading.

Reply:
> The bindings generator only checks conflicting methods of compat methods, so if it's not registered as a _compat method_ it won't check. We assume there are no conflicts in the non-compat methods because ClassDB doesn't support method overloading.

Can't there be conflicts between a compat method and a non-compat method though, like we have here apparently?

Reply:
So I've built locally to check the generated C# and I see 2 `GetMipmapOffset` methods with identical signatures:

```csharp
public long GetMipmapOffset(int mipmap)
```

Note how both methods are returning `long` (`int64_t`). So it looks like the method that returns a 64-bit integer is registered twice. Also, neither of the 2 methods is generated with the EB never attribute, this attribute is always added to compat methods:

https://github.com/godotengine/godot/blob/9508a52e2edd2aca59806e8bbba20408c36c354a/modules/mono/editor/bindings_generator.cpp#L3908-L3909

This looks like the method is registered twice as a non-compat method. So I suspect registering the compat method in ClassDB is getting confused because the method signature is the same, since the only difference in the return type is metadata.

Reply:
> I'm pretty sure we already have cases of compat methods where the only thing that changes is the return type, so this should already be handled. Not sure why it would generate invalid C# in this case, it may have something to do with the type, `int32_t` and `int64_t` are really both `INT` with different metadata, so maybe ClassDB gets confused about this and it's not properly registering the compat method.

At least for the GDExtension API, it seems to handle `int32` and `int64` properly via some metadata:
```
				{
					"name": "get_mipmap_offset",
					"is_const": true,
					"is_vararg": false,
					"is_static": false,
					"is_virtual": false,
					"hash": 923996154,
					"hash_compatibility": [
						923996154
					],
					"return_value": {
						"type": "int",
						"meta": "int64"
					},
					"arguments": [
						{
							"name": "mipmap",
							"type": "int",
							"meta": "int32"
						}
					]
				},
```

We're maybe missing handling the `meta` when the type is `Variant::INT` to handle this and recognize that it's two different signatures?

Reply:
The `hash` and the `hash_compatibility` are the same (`923996154`). @dsnopek Are you sure registering a compat method in this case is correct? I think I remember something about not adding compat methods when the only thing that changes is the types metadata.

Reply:
> The `hash` and the `hash_compatibility` are the same (`923996154`). @dsnopek Are you sure registering a compat method in this case is correct?

Ah, no, actually I think you're right. GDExtension always uses `int64_t` in its encoding anyway, so the metadata change only affects the data type that bindings generate. So, if the hash doesn't change, then we don't need the compatibility method. Sorry for the incorrect info earlier!

Reply:
So it's our CI script that needs fixing I guess?

Or should I just include the line in the `.expected` file and write that it doesn't require a compat binding?

Reply:
What is the error line? Is it actually complaining about a missing hash?

This close to 4.3, I'd say adding the line to `.expected` is probably fine for now - we can work on CI more later.

Reply:
Hm no it didn't actually complain about the hash, only:
```
Validate extension JSON: Error: Field 'classes/Image/methods/get_mipmap_offset/return_value': meta changed value in new API, from "int32" to "int64".
```

I'll silence it.

After 4.3 we could review the CI script, maybe it shouldn't report changes to `meta` at all?

Reply:
Only the errors about missing hashes really affect GDExtension binary compatibility, and call out the need for a compatibility method.

The rest are good to consider for source compatibility, like, we don't want to completely change an API signature, better to add new arguments at the end with defaults. And these errors help to call out API changes, that we might otherwise miss.

But we could certainly make the CI more clear that the hash change ones that require compatibility methods, and should only very, very rarely be ignored.